zebra-network/src/protocol/external/addr/v2.rs

```diff
@@ -0,0 +1,316 @@
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+#[error("can not parse this addrv2 variant: unimplemented or unrecognised AddrV2 network ID")]
+pub struct AddrV2UnimplementedError;
+
+impl TryFrom<AddrV2> for MetaAddr {
+    type Error = AddrV2UnimplementedError;
+
+    fn try_from(addr: AddrV2) -> Result<MetaAddr, AddrV2UnimplementedError> {
+        if let AddrV2::IpAddr {
+            untrusted_last_seen,
+            untrusted_services,
+            ip,
+            port,
+        } = addr
+        {
+            let addr = SocketAddr::new(ip, port);
+
+            Ok(MetaAddr::new_gossiped_meta_addr(
+                addr,
+                untrusted_services,
+                untrusted_last_seen,
+            ))
+        } else {
+            Err(AddrV2UnimplementedError)
+        }
+    }
+}
+
+// Just serialize in the tests for now.
+//
+// See the detailed note about ZIP-155 activation above.
+#[cfg(test)]
+impl ZcashSerialize for AddrV2 {
+    fn zcash_serialize<W: Write>(&self, mut writer: W) -> Result<(), std::io::Error> {
+        if let AddrV2::IpAddr {
+            untrusted_last_seen,
+            untrusted_services,
+            ip,
+            port,
+        } = self
+        {
+            // > uint32  Time that this node was last seen as connected to the network.
+            untrusted_last_seen.zcash_serialize(&mut writer)?;
+
+            // > Service bits. A CompactSize-encoded bit field that is 64 bits wide.
+            let untrusted_services: CompactSize64 = untrusted_services.bits().into();
+            untrusted_services.zcash_serialize(&mut writer)?;
+
+            match ip {
+                IpAddr::V4(ip) => {
+                    // > Network identifier. An 8-bit value that specifies which network is addressed.
+                    writer.write_u8(ADDR_V2_IPV4_NETWORK_ID)?;
+
+                    // > The IPV4 and IPV6 network IDs use addresses encoded in the usual way
+                    // > for binary IPv4 and IPv6 addresses in network byte order (big endian).
+                    let ip: [u8; ADDR_V2_IPV4_ADDR_SIZE] = ip.octets();
+                    // > CompactSize      The length in bytes of addr.
+                    // > uint8[sizeAddr]  Network address. The interpretation depends on networkID.
+                    zcash_serialize_bytes(&ip.to_vec(), &mut writer)?;
+
+                    // > uint16  Network port. If not relevant for the network this MUST be 0.
+                    writer.write_u16::<BigEndian>(*port)?;
+                }
+                IpAddr::V6(ip) => {
+                    writer.write_u8(ADDR_V2_IPV6_NETWORK_ID)?;
+
+                    let ip: [u8; ADDR_V2_IPV6_ADDR_SIZE] = ip.octets();
+                    zcash_serialize_bytes(&ip.to_vec(), &mut writer)?;
+
+                    writer.write_u16::<BigEndian>(*port)?;
+                }
+            }
+        } else {
+            panic!("unexpected AddrV2 variant: {:?}", self);
```

Comment:
Nit: maybe use `unreachable!` for a tiny bit more semantics?

```suggestion
            unreachable!("unexpected AddrV2 variant: {:?}", self);
```

Reply:
Technically that code is reachable by mistake, particularly from proptests.
But you're right, it's intended to be unreachable.

Fixed in c2653838b.